src/main/java/org/wso2/carbon/apimgt/securityenforcer/publisher/async/AsyncPublisher.java

```diff
@@ -0,0 +1,155 @@
+                        + requestMetaData.toString() + " as " + aseResponseDTO.getResponseMessage()
+                        + " with the response code " + aseResponseDTO.getResponseCode());
+            }
+
+            if (AISecurityHandlerConstants.ASE_RESPONSE_CODE_SUCCESS == aseResponseDTO.getResponseCode()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Access granted by the Ping AI handler for the request id " + requestCorrelationID);
+                }
+                return true;
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Access revoked by the Ping AI handler for the request id " + requestCorrelationID);
+                }
+                throw new AISecurityException(AISecurityException.ACCESS_REVOKED,
+                        AISecurityException.ACCESS_REVOKED_MESSAGE);
+            }
+        }
+    }
+
+    @Override
+    public boolean publishResponse(JSONObject requestMetaData, String requestCorrelationID) throws AISecurityException {
+        publishAsyncEvent(requestMetaData, requestCorrelationID, AISecurityHandlerConstants.ASE_RESOURCE_RESPONSE);
+        return true;
+    }
+
+    /**
+     * This class will act as thread pool executor and after executing each thread it will return runnable
+     * object back to pool. This implementation specifically used to minimize number of objects created during
+     * runtime. In this queuing strategy the submitted task will wait in the queue if the core Pool size threads are
+     * busy and the task will be allocated if any of the threads become idle.Thus ThreadPool will always have number
+     * of threads running  as mentioned in the corePoolSize.
+     * LinkedBlockingQueue without the capacity can be used for this queuing strategy.If the corePoolsize of the
+     * thread pool is less and there are more number of time consuming task were submitted,there is more possibility
+     * that the task has to wait in the queue for more time before it is run by any of the ideal thread.
+     * So tuning core pool size is something we need to tune properly.
+     * Also no task will be rejected in Thread pool until the thread pool was shutdown.
+     */
+
+    private class AsyncPublisherThreadPoolExecutor extends ThreadPoolExecutor {
+
+        AsyncPublisherThreadPoolExecutor(int corePoolSize, int maximumPoolSize, long keepAliveTime, TimeUnit unit,
+                LinkedBlockingDeque<Runnable> workQueue) {
+            super(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue);
+        }
+
+        protected void afterExecute(java.lang.Runnable r, java.lang.Throwable t) {
+            try {
+                AsyncPublishingAgent agent = (AsyncPublishingAgent) r;
+                asyncPublisherThreadPool.release(agent);
+            } catch (Exception e) {
+                log.error("Error while returning Ping AI data publishing agent back to pool" + e.getMessage());
```

Comment:
The entire Exception needs to be logged. Only the exception message is being logged currently.